### ISSUE TITLE
Add one more log level (notice/verbose?)

### ISSUE THREAD

**Issue by intgr** [rust-lang-nursery/log#334]
When writing command line programs, I often find myself wanting two distinct "info"-like log levels: one for regular output and one when a `--verbose` flag is requested by the user. Then implementing `--verbose` is simply a matter of changing the log level. I think the `Debug` level is wrong for this purpose because it's oriented towards developers, whereas `--verbose` is typically intended for troubleshooting by users.

There would be two ways to go about it:
* `Warn > Notice > Info > Debug`
* `Warn > Info > Verbose > Debug`

I realize that similar things have been requested before (#240, #229), but I think my use case is actually a quite helpful one.

If your consideration is to keep the number of built-in log levels low, I actually don't find a distinct `Trace` level to be useful; seems like debug output to me. But perhaps other people do? What was the justification for that in the first place? :)

**Comment by KodrAus** [comment]
Hi @intgr :wave:

This doesn’t sound unreasonable, however I would still argue against another info-like level for this purpose, because the levels are really arbitrary already any changes aren’t necessary going to result in quality filtering from libraries.

If you’re looking for a way to do nice optional verbose output for your own CLI then you could define whatever semantics you need for the log levels you use (as an example, using something like `env_logger`’s filtering to make `—verbose` mean logging all events raised by your CLI, but only info or higher level events from dependencies).

**Comment by sfackler** [comment]
Trace is used often in low-level libraries like hyper and tokio for very verbose logging information that can be useful to diagnose issues.

**Comment by TyPR124** [comment]
I think a "Notice" log level and a "Critical" log level would both be very nice to have. I would argue "Critical" in particular is a very necessary addition.

I was recently integrating logging for an application with Google Cloud. They define several more log levels that log does: https://cloud.google.com/logging/docs/reference/v2/rest/v2/LogEntry#logseverity

Personally I am thinking that (to an extent) the more levels `log` offers, the better. Libraries that wrap some external logging API can choose how they want to map `log`'s levels to the API, however `log` doesn't currently have enough levels to meaningfully make use of the levels provided in Google's API (or any similar API with more levels than `log` provides).

Google says that "Info" is general runtime info, such as performance stats, while "Notice" would be more significant events, like process startup, shutdown, or config changes. I think this is a meaningful difference that is worth supporting.

In terms of a "Critical" event class, I think this is especially useful (much more than "Notice") for differentiating between "this is an error that by itself isn't a big deal" and "this is an error that by itself is likely impacting my production environment negatively right now". Google offers a few more levels worse than Critical, though I I don't see those as quite as necessary in general (but, again, more levels means logging libraries can have a more fine-grained mapping of levels, and it's not like you are forced to use every log level in your application if you don't want/need them).

I don't think either Notice or especially Critical log levels are that uncommon. I know (for example) Windows supports a Critical log level.

Given this is (AFAIK) the de-facto logging facade for Rust, it is currently disappointing that I can't utilize these additional levels in Google's or anyone else's API.

**Comment by sfackler** [comment]
Adding a log level is a breaking change. The existing levels match slf4j (with the exception of a fatal level for historical reasons).

**Comment by TyPR124** [comment]
> Adding a log level is a breaking change.

Breaking changes happen. The de-factor logging crate shouldn't hinder itself by refusing to make breaking changes.

> The existing levels match slf4j

The existing levels clearly do NOT match many other logging APIs. I don't see why a single Java library is considered the authority on logging levels. Why not Google? Or Microsoft? Or Amazon? All three of these entities provide a "Critical" (or "Fatal") log level that is impossible to map from `log` in a generic and meaningful way. Surely having the ability to better support these platforms is worthwhile.

I don't mean to sound hostile (I probably do, sorry), but I don't see either as valid reasons to avoid change.

**Comment by ijackson** [comment]
I also agree that log should have both Notice and Critical.  Unix syslog has them and they are very useful.  I'm currently writing an app where I'm really feeling the lack of both of these.

**Comment by Thomasdezeeuw** [comment]
Any more movement on this? I could also use some more severity levels.

**Comment by rudolphfroger** [comment]
I think the log crate needs to support the default eight Unix syslog severity levels. See RFC3164 and RFC 5424. Many systems have used these levels as a standard to build upon and/or actually use syslog.

**Comment by shutton** [comment]
Minimally, please add 'notice'.  It's a key level in enterprise applications that's necessary for situations where you want low-noise logs, but need to see certain interesting events.  'Warning' is too strong, meriting attention, whereas 'notice' is something you'd log in case you needed to look back for something that might hint at a problem, but not until you notice something is awry.

For others, here's the overlap between syslog and log:
```
log      syslog
-------  ---------
-        emergency
-        alert
-        critical
Error    error
Warn     warning
-        notice
Info     info
Debug    debug
Trace    -
```

**Comment by jo-so** [comment]
Having a level between *Warn* and *Info* would allow addressing messages
* like *user login/logout, account creation, device came online,* ... for normal service inspection – *Notice*
* like *user saved file, user started job* ... for deeper service inspection (i.e. up-on request) – *Info*

Having a level higher than *Error* would allow distinguishing messages
* of graceful handled errors (»db connection lost«) ­– *Warn*
* of unhandled errors (»failed to write file«) – *Error*
* of errors they render the service unusable (panic, »failed to listen to socket«) – *Critical*

**Comment by KodrAus** [comment]
In general, I would argue against a proliferation of extra log levels in favor of better filtering on other properties of the log event. Levels are at best a coarse-grained proxy for the likelihood of an event being important to surface/retain. The crate currently offers a reasonable coarse-grained set of levels that cover the majority of cases. Adding more of them doesn't necessarily result in better diagnostics, and having to map them one-to-one to the levels of any other system isn't necessarily a path to a useful set of levels in `log`.

We've been steadily working towards better structured logging support over the last few years (there is also the `tracing` crate that already supports it) that would better serve users that want to work with some custom log levels, such as an associated syslog level.

**Comment by Sfinx** [comment]
Sad that there is no custom levels - I'd like to use the:

- 'fatal' level for critical messages
- 'always' level for banners that must be printed with any verbosity level set (app version data or licensing for ex.)

**Comment by ta32** [comment]
I wish there as an always level. Sometimes you want to print some stuff like what port and address a web server is running on.  I just use warning but it looks kind of weird... 

I suppose you could just make a crate and prints with the same format used by the logging system.

**Comment by KodrAus** [comment]
I think communicating end-user messages is really orthogonal to `log`, which is targeted at collecting diagnostic output. There might be some overlapping technical concerns but the details of end-user rendering I think are different enough that a dedicated domain-specific implementation would be better.

**Comment by Billy-Sheppard** [comment]
I would also love to see a new level added.

External crates especially love to fill up `INFO` with lots of useful messaging but sometimes I find wanting a `MSG` or `NOTICE` level for targeted important messages that don't make sense in `ERROR/WARN`. It is possible to get around this by using crate level log level filters but that's clunky imo, and I often want my scope all encompassing.

Custom levels would do some to alleviate this, I haven't looked into it, but would a PR be welcome?

`CRITICAL`, `IMPORTANT`, `ALERT` all feel like they could be useful.

**Comment by BarronKane** [comment]
I too would like to see a higher logging level. I've always used `FATAL` in every language as a hint that this is a potentially irrecoverable error, and in some instances I will set a `FATAL` log to trigger an automatic panic or shutdown with the log message. 

I understand this crate also strives to be compatible in `no_std` environments, but perhaps an std feature to set dynamic log levels? I've forked the repo to add in `FATAL` and have played with patching my fork in other logging libraries to some success.

Any further input I have would be a rehash of the above, but I would like to say that the existing levels truly are fine for serving crude categories of information and simply find a more critical category would be apt. My personal opinion would lean into a fully abstract implementation of all of these to be defined by the consumer (with the existing levels as default), but that would constitute an entirely new paradigm that probably necessitates a separate crate entirely. I may end up exploring this idea using this crate as a base, but the existing crates, like tracing, implement much of what I want even if they're a bit heavy.

As an aside, for the state this crate is in, I agree with @Billy-Sheppard whether `FATAL` or `CRITICAL` etc is used.

**Comment by skypexu** [comment]
Classic log levels can never meet a wide range of needs. For example, Ceph has  more granularity, like log level from 1 - 100，they are just number.  I don't think as a generical log crate, the classic log levels is enough.

**Comment by ewoolsey** [comment]
I would also like to have a `notice` and `critical` level. I don't think better filtering is a substitute for more log levels. I think following the same spec that unix does seems reasonable to me.

**Comment by cscherrNT** [comment]
To not have a breaking change, it might be a good idea to implement the new/custom log levels behind a non-default feature. This would ensure that existing crates making use of `log` have no breaking changes (for example when `match`ing the log level).

This way, users can select if they want the existing 5 log levels, or additional. Personally, I think more log levels are useful to distinguish events further. `panic!` is not really a replacement for `FATAL`, as it does not log but instantly stop the application.

**Comment by Sajjon** [comment]
[Apple's swift/log has](https://github.com/apple/swift-log):

- trace
- debug
- info
- notice
- warning
- error
- critical

Coming from Swift world, I miss `critical` and `notice`.

**Comment by MichalLebeda** [comment]
I still fail to understand what is the reasoning behind not adding more levels except it being "breaking change" which I believe would be addressed pretty quickly & easily in the impl. crates.

I understand that some users may find the current set of levels sufficient, but that may not be the case for others, especially in case of mapping levels from other systems as the bijection is often impossible.

I think the best way to figure this out would be a simple poll.

**Comment by Thomasdezeeuw** [comment]
> I still fail to understand what is the reasoning behind not adding more levels except it being "breaking change" which I believe would be addressed pretty quickly & easily in the impl. crates.

You vastly underestimating how much effort it takes to update an entire ecosystem to a new version of a crate like log.

Take a look at the downloads of Log v0.3: https://crates.io/crates/log/0.3.9, it still gets 10,000-20,000 downloads a day (in the last 90 days). If you look at it's dependents it still list 16k crates on crates.io: https://crates.io/crates/log/reverse_dependencies.

v0.4 was related 6 years ago.

6 years and it's still be used. And that's not even the main problem. The main problem is (in)compatibility between crates using log. A lot of effort was put into the v0.3 -> v0.4 transition to make it worse as smoothly a possible, but it's not perfect, same for a possible transition from v0.4 to v0.5 (or v1?).

I don't think we should be main breaking changes lightly, it creates a lot of churn in the ecosystem.

> I understand that some users may find the current set of levels sufficient, but that may not be the case for others, especially in case of mapping levels from other systems as the bijection is often impossible.

Question is where you draw the line. What levels should be included, what shouldn't and why?

**Comment by narodnik** [comment]
We definitely need a log level between Info and Debug. Right now in a large project, we have a huge trace amount of trace output which is disabled by default. But then even debug is seriously noisy. We have to maintain a large list of targets that need muting for debug to be usable. We are abusing the targets when ideally there would be an extra log level.

**Comment by kellytk** [comment]
As it's been six years, movement or maintainer comment would be greatly appreciated.

**Comment by Thomasdezeeuw** [comment]
> As it's been six years, movement or maintainer comment would be greatly appreciated.

Both @KodrAus and myself made comments on this thread pointing out issues, has there been any progress on fixing these?

**Comment by tisonkun** [comment]
One issue is that we expose `Level`'s and `LevelFilter`'s repr as usize, and there is no space between each level.

It should be possible to leave some space between levels, such as:

```
Error = 300
Warn = 400
Info = 500
Debug = 600
Trace = 700
```

Then when we want to add more level like NOTICE/CRITICAL, we won't break ourselves.

Users should not depend on the repr but use them as enum. Ord and Eq still works, but no prev/next.

Refs -

* https://github.com/apache/logging-log4j2/blob/9b9a0d43/log4j-api/src/main/java/org/apache/logging/log4j/spi/StandardLevel.java
* https://github.com/apache/logging-log4j2/blob/9b9a0d43/log4j-api/src/main/java/org/apache/logging/log4j/Level.java
* https://github.com/qos-ch/slf4j/blob/183aaa50/slf4j-api/src/main/java/org/slf4j/event/EventConstants.java
* https://github.com/qos-ch/slf4j/blob/183aaa50/slf4j-api/src/main/java/org/slf4j/event/Level.java

**Comment by KodrAus** [comment]
I've ended up softening on this issue over the years, and when we do another breaking release (which isn't planned at this stage) I'm on board with adding a level after error for critical/fatal conditions, but not on adding another notice/informational level, which we already have three of.

**Comment by alextes** [comment]
i'm in favor of "more than current". i think this crate should make breaking changes extremely rarely. like maybe the next set should be considered the final set forever and always.

what i don't understand however, beyond the stable default set, is if so many clearly want a different set and it's v hard to converge, clearly different users have different needs. why not make it configurable?

wise lib maintainers will stick with the default. wise app-space users will add whatever levels are convenient.

i see this suggestion was made back in aug 2022 already. is there an obvious reason we can't do this? i believe Rust's third core tenet was "productivity". i appreciate the effort made to find a good default set but agree with maintainers its discussion will likely last forever. but at least for the vast majority of rust users, in app space, it doesn't have to. happy to create a new issue for this.

**Comment by Thomasdezeeuw** [comment]
@alextes how do you propose to make it configurable?

**Comment by Thomasdezeeuw** [comment]
Your suggesting as a solution for this issue so I think place is fine to discuss it. If you want to create a new issue that's also fine.

I'm not going to read an AI answer though, I have no interest in that. Please consider other people time and don't waste it. If you have a though through proposal that's great, but letting AI generate something isn't that.

**Comment by psztoch** [comment]
We shouldn't blindly look only at syslog. But it certainly has additional levels that we don't have: CRITICAL (something worse than ERROR) and NOTICE (something between WARNING and INFO).
And in some applications, these additional levels are very necessary. I believe they should be added as soon as possible. They can always be flattened if the log destination doesn't support our granularity.

**Comment by kanru** [comment]
If a breaking change will be introduced, couldn't it use numerical levels and add utility to clamp values to predefined levels?

```
pub struct Level(u8);
```

For example, standard levels are modeled after syslog with 10 steps in between. 0=Critical, 10=Error, 20=Warning, etc.

Or it can use 100 steps if the repr can be larger than u8.

Numerical levels are pretty common and it allows  customization.

It might even be possible to make it ABI compatible with the current version. Just reserve the already used numbers.

**Comment by Sajjon** [comment]
> If a breaking change will be introduced, couldn't it use numerical levels and add utility to clamp values to predefined levels?
> 
> ```
> pub struct Level(u8);
> ```
> 
> For example, standard levels are modeled after syslog with 10 steps in between. 0=Critical, 10=Error, 20=Warning, etc.
> 
> Or it can use 100 steps if the repr can be larger than u8.
> 
> Numerical levels are pretty common and it allows  customization.
> 
> It might even be possible to make it ABI compatible with the current version. Just reserve the already used numbers.

I was thinking the exact same thing. Using steps of 10 should be enough. E.g:
```
Trace = 10,
Debug = 20,
Info = 30,
Notice = 40,
Warn = 50,
Error = 60,
Critical = 70,
```

**Comment by BurntSushi** [comment]
[That's similar to what Go's `log/slog` package does.](https://pkg.go.dev/log/slog#Level)

**Comment by Thomasdezeeuw** [comment]
One problem I've had in the past with numerical log levels is that two systems can disagree what level a numerical value actually is, e.g. what does `30` mean? It makes string options a lot safer as no (sane) system would parse `"critical"` as `"info"` if it didn't have a `critical` log level (like log currently). However, if we go with (public) numerical values we lose the option to convert it to string in all cases.

**Comment by KodrAus** [comment]
I would also not want to introduce numerical levels. Log levels are already z-indexes, but that’s largely offset by their semantics and coarse granularity. I don’t want to open the door to discussions like _should this log be a 102 or a 103_. This also leaves the problem of components needing to agree on what numbers any levels between levels, such as Notice, mean if they want to interpret them.

**Comment by KodrAus** [comment]
Having said that, if we were to introduce numerical log levels though, I think we could adopt the same scheme as OpenTelemetry: https://opentelemetry.io/docs/specs/otel/logs/data-model/#field-severitynumber

That leaves just a few values per level for intermediate severities, and is already compatible with our current set of levels (plus a critical/fatal level).

**Comment by SteveLauC** [comment]
IMHO, log levels to use and their semantics should be specific to applications, and crates like `log` should provide a sufficient set of levels so that applications can choose the ones they want to use.

**Comment by filmil** [comment]
In case it is helpful, perhaps one could take a look at the "glog"  API, e.g. as implemented by https://abseil.io/docs/cpp/guides/logging or https://github.com/golang/glog.

There are named log levels (FATAL, ERROR, WARNING, INFO, DEBUG), but also "verbose" logs which use an integer to denote verbosity. To avoid misunderstandings about what verbose log levels mean, verbose logging can be turned off or on per module, as well as for everything. (I also think that named log levels are given negative log level values, while verbose logs ended up being positive, but that's a detail.)

I personally found this approach rather useful over the years: have a few commonly-agreed upon logging levels, and then a large sea of "verbose" logs with arbitrary level of detail. This avoids needing to litigate the log levels that something should be logged at.  It also avoids needing to agree on the "correct" meaning of log levels across modules or programs. And in the end allows the user to provide debug information with varying levels of granularity, when needed.

**Comment by VinnieThePooh** [comment]
Fatal/Critical log level is a must here.
Coming from .NET ecosystem, i don't get how that could be an oversight for such sort of essential log level.
Regarding automation of business processes fatal/critical log level usually means some unrecoverable error that requires immediate SRE/System Administrator intervention (event itself might be extremely expensive for business).

Currently it looks like the basic logging facade is  designed not accurately enough (i would say even carelessly).

Just for reference how it looks in .NET ecosystem.
[LogLevel from MS Logging Facade](https://github.com/dotnet/runtime/blob/0fb17f7320fbaaa93e91c4c08b3a03175bb5ca15/src/libraries/Microsoft.Extensions.Logging.Abstractions/src/LogLevel.cs#L44) (please note the `Critical` log level):
```
namespace Microsoft.Extensions.Logging
{
    /// <summary>
    /// Defines logging severity levels.
    /// </summary>
    public enum LogLevel
    {
        /// <summary>
        /// Logs that contain the most detailed messages. These messages may contain sensitive application data.
        /// These messages are disabled by default and should never be enabled in a production environment.
        /// </summary>
        Trace = 0,

        /// <summary>
        /// Logs that are used for interactive investigation during development.  These logs should primarily contain
        /// information useful for debugging and have no long-term value.
        /// </summary>
        Debug = 1,

        /// <summary>
        /// Logs that track the general flow of the application. These logs should have long-term value.
        /// </summary>
        Information = 2,

        /// <summary>
        /// Logs that highlight an abnormal or unexpected event in the application flow, but do not otherwise cause the
        /// application execution to stop.
        /// </summary>
        Warning = 3,

        /// <summary>
        /// Logs that highlight when the current flow of execution is stopped due to a failure. These should indicate a
        /// failure in the current activity, not an application-wide failure.
        /// </summary>
        Error = 4,

        /// <summary>
        /// Logs that describe an unrecoverable application or system crash, or a catastrophic failure that requires
        /// immediate attention.
        /// </summary>
        Critical = 5,

        /// <summary>
        /// Not used for writing log messages. Specifies that a logging category should not write any messages.
        /// </summary>
        None = 6,
    }
}
```

[Serilog's LogLevel enum definition](https://github.com/serilog/serilog/blob/7f3a5bfec73dba1c3b29780b85a53b0fe2f7c52b/src/Serilog/Events/LogEventLevel.cs#L55) (.NET's most popular 3rd-party logging provider, please notice the `Fatal` variant, it is aligned one-to-one with `Critical` from logging facade):
```
namespace Serilog.Events;

/// <summary>
/// Specifies the meaning and relative importance of a log event.
/// </summary>
public enum LogEventLevel
{
    /// <summary>
    /// Anything and everything you might want to know about
    /// a running block of code.
    /// </summary>
    Verbose,

    /// <summary>
    /// Internal system events that aren't necessarily
    /// observable from the outside.
    /// </summary>
    Debug,

    /// <summary>
    /// The lifeblood of operational intelligence - things
    /// happen.
    /// </summary>
    Information,

    /// <summary>
    /// Service is degraded or endangered.
    /// </summary>
    Warning,

    /// <summary>
    /// Functionality is unavailable, invariants are broken
    /// or data is lost.
    /// </summary>
    Error,

    /// <summary>
    /// If you have a pager, it goes off when one of these
    /// occurs.
    /// </summary>
    Fatal
}
```
So, i am voting for fatal variant. 
Looking at previous comments the majority of different stacks have well aligned logging levels in their ecosystem, except of Rust.
